lib/ansible/modules/async_wrapper.py

```diff
@@ -241,7 +241,7 @@ def main():
     async_dir = os.environ.get('ANSIBLE_ASYNC_DIR', '~/.ansible_async')
 
     # setup job output directory
-    jobdir = os.path.expanduser(async_dir)
+    jobdir = os.path.expanduser(os.path.expandvars(async_dir))
```

Comment:
Variable substitution is handled automatically for plugin option types like `path`, `pathspec`, and `pathlist`. You can fix this by removing the code change and documenting `type: path` for this option: https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/doc_fragments/shell_common.py#L52. Also, it seems like `remote_tmp` should have the same type.